### ISSUE TITLE
Would it be possible to share the pre-trained VGG-16 model? 

### ISSUE THREAD

**Issue by ProGamerGov** [rgeirhos/texture-vs-shape#9]
I am really interested in using a stylized VGG-16 model with [neural-style-pt](https://github.com/ProGamerGov/neural-style-pt) (a pytorch version of the original [neural-style](https://github.com/jcjohnson/neural-style)), as the style transfer results show in https://github.com/rgeirhos/Stylized-ImageNet look really interesting.

I don't have access to the resources to train a full VGG-16 model with the Stylized-ImageNet dataset.

**Comment by rgeirhos** [comment]
Implemented in commit 4053aa364e97abfc490298f89ae96f8ba05c000c.
Please let me know how style transfer with a SIN-trained VGG goes! ;)

**Comment by ProGamerGov** [comment]
Thanks for sharing the model!

So, I used this script to make the model more compatible with neural-style-pt:

```
import torch
from collections import OrderedDict

state_dict = torch.load('master/vgg16_train_60_epochs_lr0.01-6c6fcc9f.pth.tar')
state_dict = state_dict['state_dict']

new_state_dict = OrderedDict()
for k, v in state_dict.items():
            if 'module.' in k:
                name = k.replace('module.', '') # remove `module.`
            else:
                name = k
            new_state_dict[name] = v

torch.save(new_state_dict, 'vgg-16-fixed-imagenet-stylized.pth')
```

I modified neural-style-pt to use the correct image pixel value range and mean/std stuff: https://github.com/ProGamerGov/neural-style-pt/blob/stylized-imagenet/neural_style.py#L328-L361

```
python3 neural_style.py -backend cudnn -model_file vgg-16-fixed-imagenet-stylized.pth -seed 876 -style_weight 10000
```

![out](https://user-images.githubusercontent.com/10626398/67051687-09982680-f0f9-11e9-96db-55a3d137c36b.png)

There looks to be some sort of instability or something like that with the results (the gray haze). I'm not sure if that has something to with normalizing activations, like in the original neural-style with the original normalized VGG model shared by Bethge Lab?

**Comment by ProGamerGov** [comment]
I borrowed some code from [NeuralStyleTransfer](https://github.com/leongatys/PytorchNeuralStyleTransfer/blob/master/NeuralStyleTransfer.ipynb) to test normalizing the weights:

```
    style_weight_list = [100000/n for n in [64,128,256,512,512]]
    for n, i in enumerate(style_losses): 
        i.strength = style_weight_list[n] * 100
```

I used a `-content_weight` of `1`, and the stylized imagenet VGG model: 

![out](https://user-images.githubusercontent.com/10626398/67054616-61875b00-f102-11e9-8b8f-967b087c3429.png)


This looks like it might be the right way forward, but I had to use some really high weight numbers to make it work.


![out](https://user-images.githubusercontent.com/10626398/67061222-e2ebe700-f11c-11e9-879e-c507a121c0a5.png)

Zero content weight: 

![out](https://user-images.githubusercontent.com/10626398/67061701-4f1b1a80-f11e-11e9-822c-173ab2e3a4a8.png)

**Comment by rgeirhos** [comment]
Thanks for sharing, @ProGamerGov, I appreciate it! :)

So your conclusion would be that style transfer with SIN-trained VGG-16 works, but needs much bigger style weights?

**Comment by ProGamerGov** [comment]
@rgeirhos Yes, I think that in order to get good results you need to use content/style weights^2 to get comparable results to the normal VGG models.  


I was able to make it so that 'normal' weights can be used with: 

```
for n, i in enumerate(module_losses): 	
     i.strength = i.strength**2 / max(i.target.size())
```

For the examples below, I generated a 512px output image first, before generating a 720px output using multiscale generation. The same input parameters were used, with only the model and whether normalization of weights was performed. You can click on the images to view them in a bigger size.

The default VGG-16 Caffe model is just the VGG-16 model from [here](https://www.robots.ox.ac.uk/~vgg/research/very_deep/), converted from Caffe to PyTorch.



**SIN-Trained VGG-16:**

Normalized content/style weights and the power of 2 operation are on the left, and on the right only the power of 2 operation was performed on the content/style weights:
<div align="center">
<img src="https://user-images.githubusercontent.com/10626398/67116834-83362000-f19e-11e9-8914-4163a3416d16.png" width="250px">
<img src="https://user-images.githubusercontent.com/10626398/67116893-9cd76780-f19e-11e9-8652-8f911768f11d.png" width="250px">
</div>

**Caffe VGG-16:**

The default Caffe VGG-16 model with normalized content/style weights and the power of 2 operation is on the left, and only the power of 2 operation was used on the left: 

<div align="center">
<img src="https://user-images.githubusercontent.com/10626398/67118444-23417880-f1a2-11e9-8197-71801b6ccf3b.png" width="250px">
<img src="https://user-images.githubusercontent.com/10626398/67117499-16bc2080-f1a0-11e9-8c41-25fa0aa9063d.png" width="250px">
</div>

And finally here is the default Caffe VGG-16 model with no normalization or power of 2 operation:

<div align="center">
<img src="https://user-images.githubusercontent.com/10626398/67119500-86cca580-f1a4-11e9-86df-1ef57e95bc18.png" height="250px">
</div>

---

The SIN-trained VGG-16 model seems to be similar in terms of parameters to the `vgg_normalised.caffemodel`:

https://github.com/jcjohnson/neural-style/issues/7
https://github.com/jcjohnson/neural-style/issues/103

**Comment by ProGamerGov** [comment]
For comparison, here's the SIN-trained VGG-16 model with no power of two operation or weight normalization, using the same settings as above.

![2_out](https://user-images.githubusercontent.com/10626398/67120358-5f76d800-f1a6-11e9-86e0-7e74997eb5f3.png)

**Comment by ProGamerGov** [comment]
These are a "heatmap" of what excites a model most on layer relu4_2, that I created with this [script](https://github.com/ProGamerGov/pytorch-convis/blob/master/convis_heatmap.py).


On the left is the Caffe VGG-16 model, and on the right is the SIN-trained VGG-16 model:
<div align="center">
<img src="https://user-images.githubusercontent.com/10626398/67129054-d1f1b300-f1ba-11e9-8d1d-bb79ae236322.png" width="250px">
<img src="https://user-images.githubusercontent.com/10626398/67163389-772b9900-f32b-11e9-9fdb-4791f99b38d6.png" width="250px">
</div>

And this is the pre-trained Torchvision model:

<div align="center">
<img src="https://user-images.githubusercontent.com/10626398/67129058-d5853a00-f1ba-11e9-8c49-52ca549515bd.png" width="250px">
</div>


It looks like the SIN-trained VGG-16 model has fewer strong activations.

**Comment by ProGamerGov** [comment]
I had mistakenly thought that the SIN-trained model did not use Caffe preprocessing, but it turns out there was also a bug that use Caffe preprocessing when I used the SIN-trained model. Sorry if there was any confusion with my previous comments (which I've now corrected). 


Also, would it be okay for me to share the SIN-trained vgg-16 on it's own, to make it easy for users of neural-style-pt to use, alongside [other models](https://github.com/ProGamerGov/neural-style-pt/wiki/Other-Models)?
